RFC: rewrite_sourceloc! function

c42f opened the pull request:
A tool to propagate the implicit `__source__` variable from a macro to any
macro calls in the generated AST.

This automates @vtjnash's trick from https://github.com/JuliaLang/julia/pull/21746#discussion_r121573359 and makes it accessible to people who want to do the common task of making uses of `@__LINE__` inside a macro expand to the location of the macro caller, rather than a location inside the macro.  (The trick is simple to use directly in some cases, but this tool makes it a *lot* easier to use if `@__LINE__` is deeply embeeded in the AST.)

An obvious question is whether this should propagate into nested macro calls when processing the AST.  For now, I've not done this but perhaps I need to think a bit more about that.

c42f committed:
Add replace_sourceloc!() to rewrite macro __source__

A tool to propagate the __source__ variable from a macro to any
macro calls in a given AST such that the inner macro calls are
attributed the same line and file name.

c42f commented on base/loading.jl:
```diff
@@ -859,3 +859,40 @@ macro __DIR__()
+function propagate_sourceloc(sourceloc, ex)
+    if ex isa Expr
+        if ex.head == :macrocall
+            ex.args[2] = sourceloc
```
Thinking about it, I'd better recurse into this as well.

vtjnash commented on base/loading.jl:
```diff
@@ -859,3 +859,40 @@ macro __DIR__()
+"""
+macro propagate_sourceloc(ex)
+    :(propagate_sourceloc($(esc(:__source__)), $(esc(ex))))
+end
```
This macro feels unnecessary to me - using the function form alone seems cleaner.

c42f replied:
Sure. How will we deal with 0.6 compatibility?

For context, I ported this utility from here: https://github.com/c42f/MicroLogging.jl/blob/e37dffcb7aeb9c581a27b897fa9d4fe3467d8b85/src/util.jl#L34 where I'm trying to support both 0.6 and 0.7; in 0.6 the `__source__` variable isn't present which is the reason for the macro.

What's the sanest way to deal with getting a nop version of this into `Compat`?  I could add something to make the following work in 0.6, by just removing the function call.  A bit messy!

```julia
@compat propagate_sourceloc(__source__, quote
    @__LINE__
end)
```

Alternatively I could just move the `@propagate_sourceloc` macro to `Compat`...

vtjnash replied:
how about:
````julia
code = quote end
Compat.macros_have_sourceloc && propagate_sourceloc!(__source__, code)
return code
```

c42f replied:
Ok then.  The few people who care about this (myself and two others?) can delete the small amount of mess once 0.7 is out - no need to make it super complicated.

c42f commented:
I changed some things as suggested and added some documentation.

tkelman commented on doc/src/stdlib/file.md:
```diff
@@ -50,6 +50,7 @@ Base.Filesystem.basename
 Base.@__FILE__
 Base.@__DIR__
 Base.@__LINE__
+Base.propagate_sourceloc!
```
doesn't seem like the right manual section for this?

c42f replied:
Correct, however this is most related to `@__LINE__`, `@__FILE__` and `@__DIR__`.  Perhaps all these should be moved to another location.  The location in the manual ("metaprogramming") seems much more sensible for these things, as they're all about introspection of the macro expansion process.

c42f replied:
How about moving all four of these to "doc/src/stdlib/base.md"?

c42f replied:
I've moved these around - hopefully into a more coherent location in the stdlib docs.

tkelman commented on base/exports.jl:
```diff
@@ -1214,6 +1214,7 @@ export
     @__FILE__,
     @__DIR__,
     @__LINE__,
+    propagate_sourceloc!,
```
does this need to be exported?

c42f replied:
Good question.  That depends on whether there's a distinction between "public API" and "exported function". I want this as a tool which can be used by packages, but it's also relatively obscure. So if there's a well defined category "public but not exported" this would fit in there, and we could not bother to export it.

tkelman replied:
There are some things that are public but not exported, usually it's whether they are documented in a public way as if to be used qualified. Most of those are under more specific modules than Base though, like Profile. Could maybe move this to the Meta module?

c42f replied:
Ok, let's try this and see how it looks.

tkelman replied:
cool, so if it's in Meta it probably shouldn't be re-exported from Base too

c42f replied:
Ah yes.  I could have sworn I fixed that.

tkelman commented on doc/src/manual/metaprogramming.md:
```diff
@@ -534,6 +534,25 @@ LineNumberNode
+julia> dump(@__LOCATION__())
+Tuple{Int64,String}
+  1: Int64 1
+  2: String "REPL[2]"
```
will this pass when you run the doctests?

c42f replied:
It will now ;-)

c42f commented:
Now moved to `Base.Meta` and exported from there.  There seemed to be a couple of useful exported but non documented functions in `Base.Meta` so I had a go at documenting them while I was at it.

c42f commented:
Looks like the travis failure is a timeout.

ihnorton commented:
Why base?

ihnorton commented on base/meta.jl:
```diff
@@ -7,17 +7,66 @@ module Meta
+macro example()
+    propagate_sourceloc!(__source__,
+        quote
+            @__LINE__
```
I think I'm not understanding the purpose of this example, so I'm wary of, erhm, propagating it... If it's trying to solve https://github.com/c42f/MicroLogging.jl/commit/f6db3715c681af90511f5219252f16e57096a194#commitcomment-20846418 then why not interpolate `__source__.line`? `@__LINE__` is documented and implemented as a literal (that's why we need the parser to insert it..), so I think this would be confusing.

c42f replied:
This is somewhat related to that old comment, but addresses the more general problem of wrapping macros which make use of `__source__`, for example here:

https://github.com/c42f/MicroLogging.jl/blob/e37dffcb7aeb9c581a27b897fa9d4fe3467d8b85/src/MicroLogging.jl#L205

If it were just `@__LINE__` I was worried about, interpolating `__source__.line` would be a better solution.

c42f replied:
That is, if package `A` provides the macro

```julia
macro a()
    quote
        some_complicated_expression($(__source__.line))
    end
end
```

And package `B` wishes to wrap it - 

```julia
macro b()
    quote
        # ...
        @a
    end
end
```

while preserving reporting of the source location, you either need a tool a bit like this, or to manually dig into the AST generated in `@b` and rewrite the source location of the macro invocations contained within.

c42f commented:
> Why base?

Because without this small tool, using `@__LINE__` and `@__FILE__` inside a macro expands to the location inside the macro, which is almost never what you want. So in this sense, it belongs in the standard library, to make some other core features of `Base` usable.

ihnorton commented:
Someone else should probably comment/approve. My thoughts are:

- not clear on why this needs to be recursive vs. a single-shot forwarding call like in the discussion thread with Jameson.
- I would prefer not exporting, and removing the example, at least from the narrative Metaprogramming section. Macros-calling-macros is fairly advanced and I don't think that such an example should be in the intro docs (at some point we may need to split some content off to an "Advanced Metaprogramming" section). Many issues such as this can be avoided by making the code generator a function.

ihnorton commented:
Generally it would be great to find a more general way to handle this issue, e.g. during expansion, but given that AFAICT [Clojure hasn't solved it either](https://dev.clojure.org/jira/browse/CLJ-865), I suspect it will be very challenging.

Also very closely related: http://tratt.net/laurie/blog/entries/debugging_layers

c42f committed:
Documentation for quot() and isexpr()

Also find an improved location for __FILE__, __DIR__ and __LINE__ in the
manual.

c42f commented:
I've just wanted this again, so I've had occasion to come back to this and clean it up as suggested (the function is exported by `Meta`, but not by `Base`, so I've left that alone).

As to whether this needs to be recursive - well obviously it doesn't, but it's a lot nicer for it to be recursive when the macro call or calls are nested deeply in the AST.

I fully agree that having shared functions which generate `Expr`s for use in multiple macros is generally nicer. This is a good idea if you have control over both macros.  But when the called macro is from a separate module this isn't so easy.

StefanKarpinski commented:
The functionality 👍, that name though... Of course, I don't have any better ideas.

c42f commented:
Yes, the name is now quite a lot worse than it was when this PR was first submitted. I think I'll change the name to `replace_sourceloc!` instead which is a lot more descriptive.

(Regarding length, both names are long but I think that's desirable for obscure and little used functionality.)

c42f commented:
Renamed.  I had a think, but I'm all out of names which describe this operation clearly but succinctly. `replace_sourceloc` is at least consistent with `Compat.macros_have_sourceloc`.

c42f commented:
I assume the OSX Travis failures are spurious.  It's hard to imagine how they could be related to the changes here.

KristofferC commented:
Could this be used to fix the following:

```
julia> @test_nowarn warn("fds")
Test Failed at /Users/kristoffer/julia/usr/share/julia/site/v0.7/Test/src/Test.jl:521
```

Here the source location info from the `@test` is printed but that is wrapped inside the `@test_nowarn´ macro and get the wrong location. We really want the location where `@test_nowarn` was expanded.

c42f commented:
Yes, this is exactly what this function is for.

(Though I'm just about to make `@test_warn` obsolete, I hope!)

vtjnash commented:
Anyone remember: why didn't we ever merge this?

c42f commented:
I don't know why this wasn't merged; I think mainly the discussion went on long enough that I eventually forgot about it.

Thanks for rebasing. I still use this trick occasionally, I just tend to write it out by hand.
